include/boost/geometry/algorithms/detail/envelope/linestring.hpp

```diff
@@ -0,0 +1,221 @@
+    static inline OutputIterator push_interval(Longitude const& lon1,
+                                               Longitude const& lon2,
+                                               OutputIterator oit)
+    {
+        typedef longitude_interval<Longitude> interval_type;
+
+        typedef math::detail::constants_on_spheroid
+            <
+                Longitude, radian
+            > constants;
+
+        BOOST_ASSERT(! math::larger(lon1, lon2));
+        BOOST_ASSERT(! math::larger(lon1, constants::max_longitude()));
+
+        if (math::larger(lon2, constants::max_longitude()))
+        {
+            *oit++ = interval_type(lon1, constants::max_longitude());
+            *oit++ = interval_type(constants::min_longitude(),
+                                   lon2 - constants::period());
+        }
+        else
+        {
+            *oit++ = interval_type(lon1, lon2);
+        }
+        return oit;
+    }
+
+    template <typename Linear, typename Box>
+    static inline void apply(Linear const& linear, Box& mbr)
+    {
+        typedef typename coordinate_type<Box>::type box_coordinate_type;
+        typedef typename fp_coordinate_type<Linear>::type calculation_type;
+
+        typedef typename geometry::segment_iterator
+            <
+                Linear const
+            > iterator_type;
+
+        typedef typename coordinate_system
+            <
+                Linear
+            >::type::units linear_units_type;
+
+        BOOST_ASSERT(geometry::num_points(linear) != 0);
+
+        typedef longitude_interval<calculation_type> interval_type;
+
+        calculation_type const pi = math::pi<calculation_type>();
+
+        std::vector<interval_type> longitude_intervals;
+        std::back_insert_iterator
+            <
+                std::vector<interval_type>
+            > oit(longitude_intervals);
+
+        calculation_type lat_min = 0, lat_max = 0;
+        bool first = true;
+        for (iterator_type seg_it = geometry::segments_begin(linear);
+             seg_it != geometry::segments_end(linear);
+             ++seg_it, first = false)
+        {
+            typename point_type<Linear>::type p[2];
+            detail::assign_point_from_index<0>(*seg_it, p[0]);
+            detail::assign_point_from_index<1>(*seg_it, p[1]);
+
+            calculation_type lon1 = geometry::get<0>(p[0]);
+            calculation_type lat1 = geometry::get<1>(p[0]);
+            calculation_type lon2 = geometry::get<0>(p[1]);
+            calculation_type lat2 = geometry::get<1>(p[1]);
+
+            math::normalize_spheroidal_coordinates
+                <
+                    linear_units_type
+                >(lon1, lat1);
+
+            math::normalize_spheroidal_coordinates
+                <
+                    linear_units_type
+                >(lon2, lat2);
+
+            math::convert_coordinates<linear_units_type, radian>(lon1, lat1);
+            math::convert_coordinates<linear_units_type, radian>(lon2, lat2);
```

Comment:
I will try to answer your comments/questions one-by-one.

*I'm guessing that this pattern can be used often. Would it be a good idea to have one well-defined function for this instead of several small ones?*

I will think about it, and yes it would make sense. I would even argue that the correct place to put all these is in `bg::convert()`.

*Why does `convert_coordinates` was added? We already have `get_as_radian`.*

I want to first normalize and then convert. I find this more robust.

*I'm guessing that `normalize_spheroidal_coordinates` are normalized using the native units (radians or degrees).*

Correct.

*Is it safe?*

Why do you think it would not be safe? What kind of safety do you refer to?

*Let's say the coordinates are normalized using degrees and then converted to radians, are the radians normalized? I'm guessing that in some cases this may not hold.*

Yes, the radians are normalized. Normalization has nothing to do with the specific units you use, and you can express it in either degrees or radians (or any other units, like seconds of degrees). Normalization basically tells you that latitudes go from `-period/4` to `+period/4` and longitudes go from `-period/2` to `period/2`. The actual value for `period` depends on the units you choose.

*I think that first the coordinates should be converted to radians and then normalized.*

Why do you think so? When you convert to radians first you immediately enter the floating-point world, even for integral coordinates. I would argue that you should delay that as long as possible.

*Furthermore I think it should be done this way in all algorithms for consistency. Therefore we probably doesn't need those generic utils you added for conversion, even more, they can be harmful. We should be forced to convert the units always the same way in all functions. Most preferably we could have single one function doing it.*

So far we have been converting to radians because we needed to feed coordinates to trigonometric functions; this is absolutely understandable. But if you do not need to use trigonometric functions why bother converting?
I do not follow you regarding the *harmful* part. I agree that we could/should encapsulate conversion into a single function, but I do not (not yet at least) get what is problematic with my approach.